File: src/Kassensystem/Kasse.java
package Kassensystem;

import java.util.Scanner;

public class Kasse {

    static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws EANException {
        System.out.println(validChecksum(scannerInput()));
    }

    static long scannerInput() throws EANException {
        System.out.print("Zahl eingeben:");
        return zahlEingeben(scanner.next());
    }

    static long zahlEingeben(String scannerInput) throws EANException {
        long ean;
        try {
            ean = Long.parseLong(scannerInput);
        } catch(NumberFormatException e) {
            throw new EANWrongDigitException();
        }

        if(!validLength(ean)) {
            throw new EANWrongLengthException();
        }
        if(!validChecksum(ean)) {
            throw new EANWrongChecksumException();
        }

        return ean;
    }

    static boolean validLength(long ean) {
        return Long.toString(ean).length() == 8 || Long.toString(ean).length() == 13;
    }

    static boolean validChecksum(long input) {
        int checksum = 0;
        String ean = Long.toString(input);
        boolean drei = Long.toString(input).length() == 8;

        for (char digit:ean.toCharArray()
             ) {
            int zahl = Integer.parseInt(Character.toString(digit));

            if(drei) {
                checksum = checksum + 3*zahl;
            } else {
                checksum = checksum + zahl;
            }
            drei=!drei;
        }
        return checksum % 10 == 0;
    }
}

File: src/KassensystemKotlin/Kasse.kt
package KassensystemKotlin

import java.util.*

@Throws(EANWrongChecksumException::class, EANException::class)
private fun main() = try {
    println(validChecksum(zahlEingeben()))
}
catch (e: EANWrongChecksumException) {
    println("Checksumme stimmt nicht")
}

private var scanner = Scanner(System.`in`)

@Throws(EANWrongChecksumException::class)
fun validChecksum(input: Long): Boolean {
    var checksum = 0
    val ean = input.toString()
    var drei = input.toString().length == 8
    for (digit in ean.toCharArray()) {
        val zahl = digit.toString().toInt()
        checksum = if (drei) {
            checksum + 3 * zahl
        } else {
            checksum + zahl
        }
        drei = !drei
    }

    if(checksum % 10 == 0) return true
    return false
}

private fun validLength(ean: Long): Boolean {
    return ean.toString().length == 8 || ean.toString().length == 13
}

@Throws(EANException::class)
private fun zahlEingeben(): Long {
    val ean: Long
    try {
        print("Zahl eingeben:")
        ean = scanner.nextLong()
        if (!validLength(ean)) {
            throw EANWrongLengthException()
        }
        if (!validChecksum(ean)) {
            throw EANWrongChecksumException()
        }
    } catch (e: InputMismatchException) {
        throw EANWrongDigitException()
    }
    return ean
}